sys/ztimer/core.c

```diff
@@ -0,0 +1,346 @@
+                clock->last = (list == &clock->list) ? NULL : list;
+            }
+
+            list->next = entry->next;
+            if (list->next) {
+                list_entry = list->next;
+                list_entry->offset += entry->offset;
+            }
+
+            /* reset the entry's next pointer so _is_set() considers it unset */
+            entry->next = NULL;
+            break;
+        }
+        list = list->next;
+    }
+}
+
+static ztimer_t *_now_next(ztimer_clock_t *clock)
+{
+    ztimer_base_t *entry = clock->list.next;
+
+    if (entry && (entry->offset == 0)) {
+        clock->list.next = entry->next;
+        if (!entry->next) {
+            clock->last = NULL;
+        }
+        return (ztimer_t*)entry;
+    }
+    else {
+        return NULL;
+    }
+}
+
+static void _ztimer_update(ztimer_clock_t *clock)
+{
+#ifdef MODULE_ZTIMER_EXTEND
+    if (clock->max_value < UINT32_MAX) {
+        if (clock->list.next) {
+            clock->ops->set(clock, _min_u32(clock->list.next->offset, clock->max_value >> 1));
+        }
+        else {
+            clock->ops->set(clock, clock->max_value >> 1);
+        }
+#else
+    if (0) {
+#endif
+    }
+    else {
+        if (clock->list.next) {
+            clock->ops->set(clock, clock->list.next->offset);
+        }
+        else {
+            clock->ops->cancel(clock);
+        }
+    }
+}
+
+void ztimer_handler(ztimer_clock_t *clock)
+{
+    DEBUG("ztimer_handler(): %p now=%"PRIu32"\n", (void *)clock, clock->ops->now(clock));
+    if (ENABLE_DEBUG) {
+        _ztimer_print(clock);
+    }
+
+#if MODULE_ZTIMER_EXTEND || MODULE_ZTIMER_NOW64
+    if (IS_USED(MODULE_ZTIMER_NOW64) || clock->max_value < UINT32_MAX) {
+        /* calling now triggers checkpointing */
+        uint32_t now = ztimer_now(clock);
+
+        if (clock->list.next) {
+            uint32_t target = clock->list.offset + clock->list.next->offset;
+            int32_t diff = (int32_t)(target - now);
+            if (diff > 0) {
+                DEBUG("ztimer_handler(): %p postponing by %"PRIi32"\n", (void *)clock, diff);
+                clock->ops->set(clock, _min_u32(diff, clock->max_value >> 1));
+                return;
+            }
+            else {
+                DEBUG("ztimer_handler(): %p diff=%"PRIi32"\n", (void *)clock, diff);
+            }
+        }
+        else {
+            DEBUG("ztimer_handler(): %p intermediate\n", (void *)clock);
+            clock->ops->set(clock, clock->max_value >> 1);
+            return;
+        }
+    }
+    else {
+        DEBUG("ztimer_handler(): no checkpointing\n");
+    }
+#endif
+
+    clock->list.offset += clock->list.next->offset;
+    clock->list.next->offset = 0;
+
+    /* ensure we're not firing too early
+     * this actually *shouldn't happen*. it does, though, when using
+     * conversions that don't round properly.
+     */
+    while ((int32_t)(clock->list.offset - ztimer_now(clock)) > 0) {}
```

Comment:
Do we really want to block during an interrupt when there are rounding issues with the conversion? I can imagine this can block for quite a while if there are only small rounding issues in `ZTIMER_SEC` to `ZTIMER_MSEC` (or from `ZTIMER_MSEC` to `ZTIMER_USEC` depending on your definition of "quite a while") conversion.

Reply:
I'm not sure on this one. There was an issue on fe310 (using 32kHz-to-1MHz-"conversion") without this.